Real compilers never crash

slavapestov opened the pull request:
Work in progress branch while master is locked.

slavapestov committed:
Sema: Remove another workaround and fuzz harder

We used to drop the entire generic parameter list if one of the
entries failed to parse. This caused a problem where the generic
parameters were still available for name lookup, so they had
to be special-cased since there's no generic environment set up
in this case.

Now, keep the parts of the generic parameter list around that
parsed successfully.

When I first made the change, almost a hundred crashers regressed;
now all the underlying issues have been fixed.

The result is that in addition to removing a crappy hack we get
some more mileage out of the compiler_crashers, because stuff like
this now builds a generic environment:

class S<T{...}

slavapestov committed:
AST: Move the well-formedness check from mangling to GenericSignature constructor

slavapestov committed:
Parser: Don't return valid location from parseGenericWhereClause() if there were no requirements

practicalswift commented:
Wow! 14 crashers fixed – the best 🎄 🎁 ever!

Hooray for @slavapestov! 🥇

practicalswift commented:
Now at 19 crashers fixed! Nice! 🥇

slavapestov commented:
@xedin and I are at 19 crashes each now :-)

xedin commented:
@slavapestov \o/ 
l will keep working on them too, maybe we can drop the count to 0 by new year :)

slavapestov commented:
Let me know which ones you're looking at so we don't duplicate work :-)

xedin commented:
I am going to take a look at what is left in your branch a bit later today, or if you have something already just let me know!

slavapestov commented:
I looked at 28586 and 28592 a bit but couldn't figure out an elegant solution. The problem is we have InOutExpr in bad places and we end up with InOutType where we don't expect one as a result. I tried just failing constraints if one of the types was not materializable, but then the next problem is if you have a type like (in_out_type (type_var_type ...)), and the type variable itself has an InOutType then we crash inside simplifyType().

slavapestov commented:
28544 looks interesting too -- infinite recursion in ConstraintSystem::solve()

slavapestov committed:
Parse: Fix

xedin commented:
Cool! I will start with 28544 and then take on other two, thanks!

xedin commented:
@slavapestov I'll take 28505 as well since this is closure + call expression...

slavapestov committed:
WIP crap

practicalswift commented:
Now at 25 crashers fixed! 👍

xedin commented:
@slavapestov 28586 and 28592 are both the good old `computeAssignDestType` related because if it sees type as type variable it creates another one and tries to bind original to it, which means that original type is "fixed" to the `@lvalue T`, I was slamming my head for a while trying to figure out how to go about it as well, `_? = ...` got fixed by unwrapping optional if destination if discarded but it was exactly the same problem...

xedin commented:
@slavapestov I'll take 28449 next while still thinking how to go about `computeAssignDestType` related problems...

slavapestov commented:
@xedin Unless you've started on any of these (I haven't, beyond root causing the issue) I'm calling dibs on 28350, 28389, 28508, 28509, 28482, and 28420. You can have the rest :-)

practicalswift commented:
@slavapestov @xedin - really amazing work from you guys over the last couple of days! Really nice to see all those crashers being fixed!

This is the current status for the swiftc crasher cleanup (SCC) of December Holiday season 2016!

Currently there are 82 crashers marked as unresolved in the repo. Of those 82 crashers ...
* 49 crashers have been fixed in submitted PRs
* 6 crashers have been claimed by @slavapestov
* 27 crashers are still up for grabs :-)

Fixed:
* 28180-rawrepresentable-extension-with-initializer.swift (fixed by @slavapestov in #6484)
* 28293-irgensilfunction-visitfullapplysite.swift (fixed by @slavapestov in #6483 and #6484)
* 28377-swift-expr-walk.swift (fixed by @slavapestov in #6484)
* 28402-swift-typebase-getcanonicaltype.swift (fixed by @xedin in 6481)
* 28413-swift-typebase-getcanonicaltype.swift (fixed by @xedin in 6481)
* 28421-swift-constraints-constraintsystem-diagnosefailureforexpr.swift (fixed by @slavapestov in #6484)
* 28424-swift-valuedecl-getformalaccessscope.swift (fixed by @slavapestov in #6484)
* 28425-swift-constraints-solution-convertbooleantypetobuiltini.swift (fixed by @slavapestov in #6484)
* 28440-swift-typechecker-resolveidentifiertype.swift (fixed by @slavapestov in #6484)
* 28449-impl-getgraphindex-typevariables-size-out-of-bounds-index-failed.swift (fixed by @xedin in 6481)
* 28470-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in 6481)
* 28473-typevariables-impl-getgraphindex-typevar-type-variable-mismatch.swift (fixed by @xedin in 6481)
* 28481-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in 6481)
* 28488-conforms-replacement-haserror-isopenedanyobject-replacement-replacement-is-gener.swift (fixed by @slavapestov in #6484)
* 28490-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28494-conforms-type-does-not-conform-to-protocol.swift (fixed by @slavapestov in #6484)
* 28497-unreachable-executed-at-swift-lib-ast-type-cpp-294.swift (fixed by @xedin in #6497)
* 28501-haderror-m-is-sourcefile-m-get-sourcefile-aststage-sourcefile-typechecked-overlo.swift (fixed by @xedin in #6489)
* 28503-iscomplete-missing-inherited-mapping-in-conformance.swift (fixed by @slavapestov in #6484)
* 28505-failed-call-arguments-did-not-match-up.swift (fixed by @xedin in #6492)
* 28507-swift-typerepr-getsourcerange-const.swift (fixed by @slavapestov in #6484)
* 28512-anonymous-namespace-traversal-visit-swift-typerepr.swift (fixed by @slavapestov in #6484)
* 28515-getmutableaddressor.swift (fixed by @slavapestov in #6484)
* 28517-getstoragekind-stored-storagekind-already-set.swift (fixed by @slavapestov in #6484)
* 28518-anonymous-namespace-verifier-walktoexprpost-swift-expr.swift (fixed by @xedin in #6489)
* 28520-args-size-fnref-getnumargumentsforfullapply-partial-application-was-throwing.swift (fixed by @xedin in #6501)
* 28527-e-gettype-isassignabletype-setting-access-kind-on-non-l-value.swift (fixed by @slavapestov in #6484)
* 28535-unreachable-executed-at-swift-lib-parse-parsedecl-cpp-610.swift (fixed by @slavapestov in #6484)
* 28537-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28541-anonymous-namespace-verifier-walktodeclpost-swift-decl.swift (fixed by @slavapestov in #6484)
* 28544-swift-type-transform-llvm-function-ref-swift-type-swift-type-const.swift (fixed by @xedin in #6491)
* 28546-anonymous-namespace-verifier-walktodeclpost-swift-decl.swift (fixed by @slavapestov in #6484)
* 28549-swift-typebase-getdesugaredtype.swift (fixed by @slavapestov in #6484)
* 28554-reftype-hastypeparameter-cannot-have-a-dependent-type-here.swift (fixed by @slavapestov in #6484)
* 28560-unreachable-executed-at-swift-lib-ast-type-cpp-1104.swift (fixed by @xedin in 6481)
* 28561-input-hastypevariable-output-hastypevariable.swift (fixed by @slavapestov in #6484)
* 28562-swift-typebase-getcanonicaltype.swift (fixed by @xedin in 6481)
* 28567-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in 6481)
* 28568-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28569-swift-declcontext-getastypeortypeextensioncontext-const.swift (fixed by @slavapestov in #6484)
* 28576-anonymous-namespace-findcapturedvars-checktype-swift-type-swift-sourceloc.swift (fixed by @xedin in 6481)
* 28577-isa-x-val-cast-ty-argument-of-incompatible-type.swift (fixed by @slavapestov in #6484)
* 28578-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28580-hastype-e-expected-type-to-have-been-set.swift (fixed by @slavapestov in #6484)
* 28581-anonymous-namespace-findcapturedvars-checktype-swift-type-swift-sourceloc.swift (fixed by @xedin in 6481)
* 28582-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28583-unreachable-executed-at-swift-lib-ast-type-cpp-1098.swift (fixed by @xedin in 6481)
* 28584-loc-isvalid.swift (fixed by @slavapestov in #6484)
* 28594-anonymous-namespace-verifier-verifychecked-swift-vardecl.swift (fixed by @xedin in #6489)

Claimed by @slavapestov:
* 28350-swift-typechecker-performtypocorrection.swift (claimed by @slavapestov in #6484)
* 28389-swift-lookupvisibledecls.swift (claimed by @slavapestov in #6484)
* 28420-swift-lowering-emitconditionalcheckedcast.swift (claimed by @slavapestov in #6484)
* 28482-hasaccessibility-accessibility-not-computed-yet.swift (claimed by @slavapestov in #6484)
* 28508-unreachable-executed-at-swift-lib-sema-csgen-cpp-2656.swift (claimed by @slavapestov in #6484)
* 28509-allowoverwrite-e-haslvalueaccesskind-l-value-access-kind-has-already-been-set.swift (claimed by @slavapestov in #6484)

Not yet fixed/claimed:
* 28368-swift-expr-propagatelvalueaccesskind.swift (up for grabs!)
* 28395-swift-expr-propagatelvalueaccesskind.swift (up for grabs!)
* 28396-swift-lowering-silgenfunction-emitclosurevalue.swift(up for grabs!)
* 28419-swift-silmodule-constructsil.swift (up for grabs!)
* 28442-swift-typebase-getrvaluetype.swift (up for grabs!)
* 28447-result-case-not-implemented-failed.swift (up for grabs!)
* 28457-unreachable-executed-at-swift-include-swift-ast-cantypevisitor-h-41.swift (up for grabs!)
* 28474-unreachable-executed-at-swift-lib-ast-type-cpp-1325.swift (up for grabs!)
* 28475-swift-typechecker-validatedecl-swift-valuedecl-bool.swift (up for grabs!)
* 28479-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift (up for grabs!)
* 28483-unreachable-executed-at-swift-lib-ast-type-cpp-1117.swift (up for grabs!)
* 28491-result-case-not-implemented.swift (up for grabs!)
* 28493-resolveidenttypecomponent-swift-typechecker-swift-declcontext-llvm-arrayref-swif.swift (up for grabs!)
* 28499-start-isvalid-end-isvalid-start-and-end-should-either-both-be-valid-or-both-be-i.swift (up for grabs!)
* 28504-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (up for grabs!)
* 28510-size-t-llvm-sys-locale-columnwidth-i-gettext-i-gettext-size.swift (up for grabs!)
* 28543-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift (up for grabs!)
* 28550-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift (up for grabs!)
* 28555-unreachable-executed-at-swift-lib-ast-type-cpp-1318.swift (up for grabs!)
* 28559-result-case-not-implemented.swift (up for grabs!)
* 28585-anonymous-namespace-verifier-walktostmtpost-swift-stmt.swift (up for grabs!)
* 28586-objectty-is-lvaluetype-objectty-is-inouttype-cannot-have-inout-or-lvalue-wrapped.swift (up for grabs!)
* 28587-child-source-range-not-contained-within-its-parent-sequence-expr-type-null.swift (up for grabs!)
* 28588-unreachable-executed-at-swift-lib-sema-csapply-cpp-5770.swift (up for grabs!)
* 28590-exprtypes-e-isequal-e-gettype-expected-type-in-map-to-be-the-same-type-in-expres.swift (up for grabs!)
* 28592-objectty-is-lvaluetype-objectty-is-inouttype-cannot-have-inout-or-lvalue-wrapped.swift (up for grabs!)
* 28593-unreachable-executed-at-swift-lib-ast-type-cpp-3771.swift (up for grabs!)

Let me know if I got anything wrong! :-)

xedin commented:
@practicalswift Thanks for awesome report, very helpful! I will double check if I have missed any crashers related to result statement checking (maybe one or two at most if any), and will keep on going through the list of unclaimed...

CodaFi commented:
@xedin If you want to automate the process, 

```
../llvm/utils/lit/lit.py -sv ../build/Ninja-DebugAssert/swift-macosx-x86_64/validation-test-macosx-x86_64/compiler_crashers/ | ./utils/resolve-crashes.py
```

CodaFi commented:
Oh, forgot to mention:

I've got 28499 and 28585 today.

xedin commented:
@CodaFi Thanks!

xedin commented:
@practicalswift Looks like there are 5 more crashers which are resolved by #6481 - 28447, 28483, 28491, 28504 and 28559 (pushed as a commit for aforementioned PR).

xedin commented:
I think most of the remaining problems are related to the `computeAssignDestType` and it's way to enforce lvalueness of the destination type...

jtbandes commented:
Looking at 28395 and 28368, these are due to UnresolvedDotExpr/UnresolvedMemberRefExpr which escape the type checker, and other code is not expecting them.  In https://github.com/apple/swift/pull/5751 I fixed something related by handling UNCHECKED_EXPR, but I wonder if that's the wrong approach in general, and instead something is supposed to be preventing these from escaping TC (but is failing to do so)?

xedin commented:
@jtbandes Yes, @rudkx is working on improvements to the constraint system which is going to prevent type variables from escaping, he mentioned that in one of the comments in #6481.

rudkx commented:
@jtbandes I'm still working on finishing things up, but I added a side map for expression types to the constraint system. We're currently still writing types directly into expressions as well, until everything is updated. After that point, we should only be writing types into expressions when we apply a final solution, which should keep the type variables from escaping.

I haven't updated the types on TypeLocs or parameters yet, but that should be done as well.

More generally it would be great to have the type checker not mutate existing ASTs but that's a much bigger change.

jtbandes commented:
Yeah, I just noticed your work on this side map. Looks great! (28490 is another one caused by escaping type vars.)

I'm guessing there's not much for an outsider to help with at this point (since you've probably got a lot of work in progress there)?

xedin commented:
@rudkx FYI, while working on one of the crashers I've noticed that if system fails only types which had type variables are marked as erroneous but other are left alone and there is some code (not exactly sure what) which is relies on that behavior, so applying only "final" types to expressions upon successful solving might degrade code completion, because if there was no solution found whole AST should be marked as erroneous or unresolved, right?

xedin commented:
crasher 28590 is related to the fact that failure diagnosis is going to modify types directly on AST while solving sub-expressions and main constraint system is not going to be made aware of that...

slavapestov committed:
Sema: Don't allow GenericTypeParamTypes to be canonicalized before depth is set

Previously we had a subtle class of bugs where if getCanonicalType()
was called on a GenericTypeParamType whose GenericTypeParamDecl
did not yet have its depth set, we would forever memoize an
incorrect canonical type for this generic parameter.

practicalswift commented:
@CodaFi @xedin Great stuff! I've added your recently fixed crashes in this status update.

This is the current status for the swiftc crasher cleanup (SCC) of December Holiday season 2016!

Currently there are 82 crashers marked as unresolved in the repo. Of those 82 crashers ...
* 60 crashers have been fixed in submitted PRs
* 6 crashers have been claimed by @slavapestov
* 16 crashers are still up for grabs :-)

Changes since last update:
* 6 crashers fixed by @CodaFi in #6504
* 5 more crashers fixed by @xedin in #6481

Fixed:
* 28180-rawrepresentable-extension-with-initializer.swift (fixed by @slavapestov in #6484)
* 28504-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in #6481)
* 28585-anonymous-namespace-verifier-walktostmtpost-swift-stmt.swift (fixed by @CodaFi in #6504)
* 28293-irgensilfunction-visitfullapplysite.swift (fixed by @slavapestov in #6483 and #6484)
* 28377-swift-expr-walk.swift (fixed by @slavapestov in #6484)
* 28396-swift-lowering-silgenfunction-emitclosurevalue.swift (fixed by @CodaFi in #6504)
* 28402-swift-typebase-getcanonicaltype.swift (fixed by @xedin in 6481)
* 28413-swift-typebase-getcanonicaltype.swift (fixed by @xedin in 6481)
* 28419-swift-silmodule-constructsil.swift (fixed by @CodaFi in #6504)
* 28421-swift-constraints-constraintsystem-diagnosefailureforexpr.swift (fixed by @slavapestov in #6484)
* 28424-swift-valuedecl-getformalaccessscope.swift (fixed by @slavapestov in #6484)
* 28425-swift-constraints-solution-convertbooleantypetobuiltini.swift (fixed by @slavapestov in #6484)
* 28440-swift-typechecker-resolveidentifiertype.swift (fixed by @slavapestov in #6484)
* 28447-result-case-not-implemented-failed.swift (fixed by @xedin in #6481)
* 28449-impl-getgraphindex-typevariables-size-out-of-bounds-index-failed.swift (fixed by @xedin in 6481)
* 28457-unreachable-executed-at-swift-include-swift-ast-cantypevisitor-h-41.swift (fixed by @CodaFi in #6504)
* 28470-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in 6481)
* 28473-typevariables-impl-getgraphindex-typevar-type-variable-mismatch.swift (fixed by @xedin in 6481)
* 28481-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in 6481)
* 28483-unreachable-executed-at-swift-lib-ast-type-cpp-1117.swift (fixed by @xedin in #6481)
* 28488-conforms-replacement-haserror-isopenedanyobject-replacement-replacement-is-gener.swift (fixed by @slavapestov in #6484)
* 28490-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28491-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28494-conforms-type-does-not-conform-to-protocol.swift (fixed by @slavapestov in #6484)
* 28497-unreachable-executed-at-swift-lib-ast-type-cpp-294.swift (fixed by @xedin in #6497)
* 28499-start-isvalid-end-isvalid-start-and-end-should-either-both-be-valid-or-both-be-i.swift (fixed by @CodaFi in #6504)
* 28501-haderror-m-is-sourcefile-m-get-sourcefile-aststage-sourcefile-typechecked-overlo.swift (fixed by @xedin in #6489)
* 28503-iscomplete-missing-inherited-mapping-in-conformance.swift (fixed by @slavapestov in #6484)
* 28505-failed-call-arguments-did-not-match-up.swift (fixed by @xedin in #6492)
* 28507-swift-typerepr-getsourcerange-const.swift (fixed by @slavapestov in #6484)
* 28512-anonymous-namespace-traversal-visit-swift-typerepr.swift (fixed by @slavapestov in #6484)
* 28515-getmutableaddressor.swift (fixed by @slavapestov in #6484)
* 28517-getstoragekind-stored-storagekind-already-set.swift (fixed by @slavapestov in #6484)
* 28518-anonymous-namespace-verifier-walktoexprpost-swift-expr.swift (fixed by @xedin in #6489)
* 28520-args-size-fnref-getnumargumentsforfullapply-partial-application-was-throwing.swift (fixed by @xedin in #6501)
* 28527-e-gettype-isassignabletype-setting-access-kind-on-non-l-value.swift (fixed by @slavapestov in #6484)
* 28535-unreachable-executed-at-swift-lib-parse-parsedecl-cpp-610.swift (fixed by @slavapestov in #6484)
* 28537-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28541-anonymous-namespace-verifier-walktodeclpost-swift-decl.swift (fixed by @slavapestov in #6484)
* 28544-swift-type-transform-llvm-function-ref-swift-type-swift-type-const.swift (fixed by @xedin in #6491)
* 28546-anonymous-namespace-verifier-walktodeclpost-swift-decl.swift (fixed by @slavapestov in #6484)
* 28549-swift-typebase-getdesugaredtype.swift (fixed by @slavapestov in #6484)
* 28554-reftype-hastypeparameter-cannot-have-a-dependent-type-here.swift (fixed by @slavapestov in #6484)
* 28559-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28560-unreachable-executed-at-swift-lib-ast-type-cpp-1104.swift (fixed by @xedin in 6481)
* 28561-input-hastypevariable-output-hastypevariable.swift (fixed by @slavapestov in #6484)
* 28562-swift-typebase-getcanonicaltype.swift (fixed by @xedin in 6481)
* 28567-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in 6481)
* 28568-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28569-swift-declcontext-getastypeortypeextensioncontext-const.swift (fixed by @slavapestov in #6484)
* 28576-anonymous-namespace-findcapturedvars-checktype-swift-type-swift-sourceloc.swift (fixed by @xedin in 6481)
* 28577-isa-x-val-cast-ty-argument-of-incompatible-type.swift (fixed by @slavapestov in #6484)
* 28578-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28580-hastype-e-expected-type-to-have-been-set.swift (fixed by @slavapestov in #6484)
* 28581-anonymous-namespace-findcapturedvars-checktype-swift-type-swift-sourceloc.swift (fixed by @xedin in 6481)
* 28582-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28583-unreachable-executed-at-swift-lib-ast-type-cpp-1098.swift (fixed by @xedin in 6481)
* 28584-loc-isvalid.swift (fixed by @slavapestov in #6484)
* 28587-child-source-range-not-contained-within-its-parent-sequence-expr-type-null.swift (fixed by @CodaFi in #6504)
* 28594-anonymous-namespace-verifier-verifychecked-swift-vardecl.swift (fixed by @xedin in #6489)

Claimed:
* 28350-swift-typechecker-performtypocorrection.swift (claimed by @slavapestov in #6484)
* 28389-swift-lookupvisibledecls.swift (claimed by @slavapestov in #6484)
* 28420-swift-lowering-emitconditionalcheckedcast.swift (claimed by @slavapestov in #6484)
* 28482-hasaccessibility-accessibility-not-computed-yet.swift (claimed by @slavapestov in #6484)
* 28508-unreachable-executed-at-swift-lib-sema-csgen-cpp-2656.swift (claimed by @slavapestov in #6484)
* 28509-allowoverwrite-e-haslvalueaccesskind-l-value-access-kind-has-already-been-set.swift (claimed by @slavapestov in #6484)

Not yet fixed/claimed:
* 28368-swift-expr-propagatelvalueaccesskind.swift (up for grabs!)
* 28395-swift-expr-propagatelvalueaccesskind.swift (up for grabs!)
* 28442-swift-typebase-getrvaluetype.swift (up for grabs!)
* 28474-unreachable-executed-at-swift-lib-ast-type-cpp-1325.swift (up for grabs!)
* 28475-swift-typechecker-validatedecl-swift-valuedecl-bool.swift (up for grabs!)
* 28479-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift (up for grabs!)
* 28493-resolveidenttypecomponent-swift-typechecker-swift-declcontext-llvm-arrayref-swif.swift (up for grabs!)
* 28510-size-t-llvm-sys-locale-columnwidth-i-gettext-i-gettext-size.swift (up for grabs!)
* 28543-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift (up for grabs!)
* 28550-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift (up for grabs!)
* 28555-unreachable-executed-at-swift-lib-ast-type-cpp-1318.swift (up for grabs!)
* 28586-objectty-is-lvaluetype-objectty-is-inouttype-cannot-have-inout-or-lvalue-wrapped.swift (up for grabs!)
* 28588-unreachable-executed-at-swift-lib-sema-csapply-cpp-5770.swift (up for grabs!)
* 28590-exprtypes-e-isequal-e-gettype-expected-type-in-map-to-be-the-same-type-in-expres.swift (up for grabs!)
* 28592-objectty-is-lvaluetype-objectty-is-inouttype-cannot-have-inout-or-lvalue-wrapped.swift (up for grabs!)
* 28593-unreachable-executed-at-swift-lib-ast-type-cpp-3771.swift (up for grabs!)

CodaFi commented:
A lot of the 14 leftover are either nondeterministic memory smashers (ASAN confirms).  Destruction of `Type`s allocated in the process of solving `ConstraintSystem`s doesn't propagate to expressions and decls that got assigned those types during the solving process.  Either we need to traverse along the locators to unset these types and nuke the little RAII cleanup struct or allocate things outside the Constraint System's arena.

@rudkx, thoughts?

xedin commented:
I'm currently working on 28592, that's related to opening generics and wrapping optional inout into lvalue type when trying to match possible type parameters, next up going to be 28586 - it's similar but different cause - `computeAssignDestType` with type variables binding to lvalue.

@practicalswift 28510 this one is going to be involved as todo suggests, i don't think we wouldn't be able to tackle it any time soon.

practicalswift commented:
@xedin Regarding 28510 - you mean this comment in `llvm/lib/Support/SourceMgr.cpp`?

```
    // FIXME: This assertion is intended to catch unintended use of multibyte
    // characters in fixits. If we decide to do this, we'll have to track
    // separate byte widths for the source and fixit lines.
```

xedin commented:
@practicalswift Yeah, that's the one.

practicalswift commented:
@xedin Could one temporary work-around be to print the multi-byte characters escaped in some way instead?

In [this particular case](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28510-size-t-llvm-sys-locale-columnwidth-i-gettext-i-gettext-size.swift) the problem seems to be `丏`, so the escaped form form could be say `U+4E0F`, `\u4E0F` or the UTF-8 hex representation `0xE4 0xB8 0x8F (e4b88f)`.

What do you think? :-)

xedin commented:
I think that's not going to be a user friendly thing, which kind of beats the purpose of fix-its, so unfortunately there is no easy work around, we have to make it proper...

slavapestov commented:
I think for that one we should just file an llvm bug?

xedin commented:
Yes, I concur. Sorry, didn't mean to imply that we have to fix it in swift :)

practicalswift commented:
@xedin Thanks! I've added your two fixed crashers in this status update.

This is the current status for the swiftc crasher cleanup (SCC) of December Holiday season 2016!

Currently there are 82 crashers marked as unresolved in the repo. Of those 82 crashers ...
* 64 crashers have been fixed in submitted PRs
* 6 crashers have been claimed by @slavapestov
* 12 crashers are still up for grabs :-)

Changes since last update:
* 2 crashers fixed by @xedin in #6510 and #6511

Not yet fixed/claimed:
* [28368-swift-expr-propagatelvalueaccesskind.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28368-swift-expr-propagatelvalueaccesskind.swift)
* [28395-swift-expr-propagatelvalueaccesskind.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28395-swift-expr-propagatelvalueaccesskind.swift)
* [28442-swift-typebase-getrvaluetype.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28442-swift-typebase-getrvaluetype.swift)
* [28474-unreachable-executed-at-swift-lib-ast-type-cpp-1325.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28474-unreachable-executed-at-swift-lib-ast-type-cpp-1325.swift)
* [28479-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28479-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift)
* [28510-size-t-llvm-sys-locale-columnwidth-i-gettext-i-gettext-size.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28510-size-t-llvm-sys-locale-columnwidth-i-gettext-i-gettext-size.swift)
* [28543-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28543-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift)
* [28550-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28550-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift)
* [28555-unreachable-executed-at-swift-lib-ast-type-cpp-1318.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28555-unreachable-executed-at-swift-lib-ast-type-cpp-1318.swift)
* [28588-unreachable-executed-at-swift-lib-sema-csapply-cpp-5770.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28588-unreachable-executed-at-swift-lib-sema-csapply-cpp-5770.swift)
* [28590-exprtypes-e-isequal-e-gettype-expected-type-in-map-to-be-the-same-type-in-expres.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28590-exprtypes-e-isequal-e-gettype-expected-type-in-map-to-be-the-same-type-in-expres.swift)
* [28593-unreachable-executed-at-swift-lib-ast-type-cpp-3771.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28593-unreachable-executed-at-swift-lib-ast-type-cpp-3771.swift)

Fixed:
* 28180-rawrepresentable-extension-with-initializer.swift (fixed by @slavapestov in #6484)
* 28293-irgensilfunction-visitfullapplysite.swift (fixed by @slavapestov in #6483 and #6484)
* 28377-swift-expr-walk.swift (fixed by @slavapestov in #6484)
* 28396-swift-lowering-silgenfunction-emitclosurevalue.swift (fixed by @CodaFi in #6504)
* 28402-swift-typebase-getcanonicaltype.swift (fixed by @xedin in 6481)
* 28413-swift-typebase-getcanonicaltype.swift (fixed by @xedin in 6481)
* 28419-swift-silmodule-constructsil.swift (fixed by @CodaFi in #6504)
* 28421-swift-constraints-constraintsystem-diagnosefailureforexpr.swift (fixed by @slavapestov in #6484)
* 28424-swift-valuedecl-getformalaccessscope.swift (fixed by @slavapestov in #6484)
* 28425-swift-constraints-solution-convertbooleantypetobuiltini.swift (fixed by @slavapestov in #6484)
* 28440-swift-typechecker-resolveidentifiertype.swift (fixed by @slavapestov in #6484)
* 28447-result-case-not-implemented-failed.swift (fixed by @xedin in #6481)
* 28449-impl-getgraphindex-typevariables-size-out-of-bounds-index-failed.swift (fixed by @xedin in 6481)
* 28457-unreachable-executed-at-swift-include-swift-ast-cantypevisitor-h-41.swift (fixed by @CodaFi in #6504)
* 28470-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in 6481)
* 28473-typevariables-impl-getgraphindex-typevar-type-variable-mismatch.swift (fixed by @xedin in 6481)
* 28475-swift-typechecker-validatedecl-swift-valuedecl-bool.swift (fixed, see #6505)
* 28481-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in 6481)
* 28483-unreachable-executed-at-swift-lib-ast-type-cpp-1117.swift (fixed by @xedin in #6481)
* 28488-conforms-replacement-haserror-isopenedanyobject-replacement-replacement-is-gener.swift (fixed by @slavapestov in #6484)
* 28490-result-case-not-implemented.swift (fixed by @xedin in 6481)
* 28491-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28493-resolveidenttypecomponent-swift-typechecker-swift-declcontext-llvm-arrayref-swif.swift (fixed, see #6505)
* 28494-conforms-type-does-not-conform-to-protocol.swift (fixed by @slavapestov in #6484)
* 28497-unreachable-executed-at-swift-lib-ast-type-cpp-294.swift (fixed by @xedin in #6497)
* 28499-start-isvalid-end-isvalid-start-and-end-should-either-both-be-valid-or-both-be-i.swift (fixed by @CodaFi in #6504)
* 28501-haderror-m-is-sourcefile-m-get-sourcefile-aststage-sourcefile-typechecked-overlo.swift (fixed by @xedin in #6489)
* 28503-iscomplete-missing-inherited-mapping-in-conformance.swift (fixed by @slavapestov in #6484)
* 28504-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in #6481)
* 28505-failed-call-arguments-did-not-match-up.swift (fixed by @xedin in #6492)
* 28507-swift-typerepr-getsourcerange-const.swift (fixed by @slavapestov in #6484)
* 28512-anonymous-namespace-traversal-visit-swift-typerepr.swift (fixed by @slavapestov in #6484)
* 28515-getmutableaddressor.swift (fixed by @slavapestov in #6484)
* 28517-getstoragekind-stored-storagekind-already-set.swift (fixed by @slavapestov in #6484)
* 28518-anonymous-namespace-verifier-walktoexprpost-swift-expr.swift (fixed by @xedin in #6489)
* 28520-args-size-fnref-getnumargumentsforfullapply-partial-application-was-throwing.swift (fixed by @xedin in #6501)
* 28527-e-gettype-isassignabletype-setting-access-kind-on-non-l-value.swift (fixed by @slavapestov in #6484)
* 28535-unreachable-executed-at-swift-lib-parse-parsedecl-cpp-610.swift (fixed by @slavapestov in #6484)
* 28537-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28541-anonymous-namespace-verifier-walktodeclpost-swift-decl.swift (fixed by @slavapestov in #6484)
* 28544-swift-type-transform-llvm-function-ref-swift-type-swift-type-const.swift (fixed by @xedin in #6491)
* 28546-anonymous-namespace-verifier-walktodeclpost-swift-decl.swift (fixed by @slavapestov in #6484)
* 28549-swift-typebase-getdesugaredtype.swift (fixed by @slavapestov in #6484)
* 28554-reftype-hastypeparameter-cannot-have-a-dependent-type-here.swift (fixed by @slavapestov in #6484)
* 28559-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28560-unreachable-executed-at-swift-lib-ast-type-cpp-1104.swift (fixed by @xedin in #6481)
* 28561-input-hastypevariable-output-hastypevariable.swift (fixed by @slavapestov in #6484)
* 28562-swift-typebase-getcanonicaltype.swift (fixed by @xedin in #6481)
* 28567-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in #6481)
* 28568-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28569-swift-declcontext-getastypeortypeextensioncontext-const.swift (fixed by @slavapestov in #6484)
* 28576-anonymous-namespace-findcapturedvars-checktype-swift-type-swift-sourceloc.swift (fixed by @xedin in #6481)
* 28577-isa-x-val-cast-ty-argument-of-incompatible-type.swift (fixed by @slavapestov in #6484)
* 28578-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28580-hastype-e-expected-type-to-have-been-set.swift (fixed by @slavapestov in #6484)
* 28581-anonymous-namespace-findcapturedvars-checktype-swift-type-swift-sourceloc.swift (fixed by @xedin in #6481)
* 28582-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28583-unreachable-executed-at-swift-lib-ast-type-cpp-1098.swift (fixed by @xedin in #6481)
* 28584-loc-isvalid.swift (fixed by @slavapestov in #6484)
* 28585-anonymous-namespace-verifier-walktostmtpost-swift-stmt.swift (fixed by @CodaFi in #6504)
* 28586-objectty-is-lvaluetype-objectty-is-inouttype-cannot-have-inout-or-lvalue-wrapped.swift (fixed by @xedin in #6511)
* 28587-child-source-range-not-contained-within-its-parent-sequence-expr-type-null.swift (fixed by @CodaFi in #6504)
* 28592-objectty-is-lvaluetype-objectty-is-inouttype-cannot-have-inout-or-lvalue-wrapped.swift (fixed by @xedin in #6510)
* 28594-anonymous-namespace-verifier-verifychecked-swift-vardecl.swift (fixed by @xedin in #6489)

Claimed:
* 28350-swift-typechecker-performtypocorrection.swift (claimed by @slavapestov in #6484)
* 28389-swift-lookupvisibledecls.swift (claimed by @slavapestov in #6484)
* 28420-swift-lowering-emitconditionalcheckedcast.swift (claimed by @slavapestov in #6484)
* 28482-hasaccessibility-accessibility-not-computed-yet.swift (claimed by @slavapestov in #6484)
* 28508-unreachable-executed-at-swift-lib-sema-csgen-cpp-2656.swift (claimed by @slavapestov in #6484)
* 28509-allowoverwrite-e-haslvalueaccesskind-l-value-access-kind-has-already-been-set.swift (claimed by @slavapestov in #6484)

slavapestov commented:
@practicalswift I think most of the ones in the unclaimed list are actually @rudkx's. The first two lvalue ones look different though. Also I'm not promising to fix the ones I "claimed" any time soon, nor do I insist that I be the one to fix them -- it's just that I figured out the root causes. If anyone wants to look at them let me know -- I'll explain the fix.

xedin commented:
@slavapestov I agree and i am going to take a look at first two, since (I think) I know what the problem is... I can also help out with the ones you claimed if you want!

xedin commented:
@practicalswift I've submitted #6514 to resolve 28368 and 28395 (as well as one of the SIL crashers).

xedin commented:
@practicalswift I've submitted the last one of this year which is going to address 28588 - #6515

practicalswift commented:
@xedin Great! Update below :-)

This is the current status for the swiftc crasher cleanup (SCC) of December Holiday season 2016!

Currently there are 82 crashers marked as unresolved in the repo. Of those 82 crashers ...
* 65 crashers have been fixed in submitted PRs which will be merged once master is unlocked (great work by this team: 34 crashers fixed by @xedin, 25 crashers fixed by @slavapestov and 6 crashers fixed by @xedin)
* [One crasher](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28510-size-t-llvm-sys-locale-columnwidth-i-gettext-i-gettext-size.swift) should probably be reported upstream to LLVM.
* Two disabled crashes marked as resolved in #6505.
* The remaining 14 crashers can be divided into two distinct sets of crashes where the members of each set share the same root cause - @rudkx's upcoming improvements to the constraint system will likely solve 8 of those and @slavapestov has identified what needs to be done for the remaining 6 crashers :-)

Fixed:
* 28180-rawrepresentable-extension-with-initializer.swift (fixed by @slavapestov in #6484)
* 28293-irgensilfunction-visitfullapplysite.swift (fixed by @slavapestov in #6483 and #6484)
* 28368-swift-expr-propagatelvalueaccesskind.swift (fixed by @xedin in #6514)
* 28377-swift-expr-walk.swift (fixed by @slavapestov in #6484)
* 28395-swift-expr-propagatelvalueaccesskind.swift (fixed by @xedin in #6514)
* 28396-swift-lowering-silgenfunction-emitclosurevalue.swift (fixed by @CodaFi in #6504)
* 28402-swift-typebase-getcanonicaltype.swift (fixed by @xedin in #6481)
* 28413-swift-typebase-getcanonicaltype.swift (fixed by @xedin in #6481)
* 28419-swift-silmodule-constructsil.swift (fixed by @CodaFi in #6504)
* 28421-swift-constraints-constraintsystem-diagnosefailureforexpr.swift (fixed by @slavapestov in #6484)
* 28424-swift-valuedecl-getformalaccessscope.swift (fixed by @slavapestov in #6484)
* 28425-swift-constraints-solution-convertbooleantypetobuiltini.swift (fixed by @slavapestov in #6484)
* 28440-swift-typechecker-resolveidentifiertype.swift (fixed by @slavapestov in #6484)
* 28447-result-case-not-implemented-failed.swift (fixed by @xedin in #6481)
* 28449-impl-getgraphindex-typevariables-size-out-of-bounds-index-failed.swift (fixed by @xedin in #6481)
* 28457-unreachable-executed-at-swift-include-swift-ast-cantypevisitor-h-41.swift (fixed by @CodaFi in #6504)
* 28470-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in #6481)
* 28473-typevariables-impl-getgraphindex-typevar-type-variable-mismatch.swift (fixed by @xedin in #6481)
* 28475-swift-typechecker-validatedecl-swift-valuedecl-bool.swift (fixed, see #6505)
* 28481-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in #6481)
* 28483-unreachable-executed-at-swift-lib-ast-type-cpp-1117.swift (fixed by @xedin in #6481)
* 28488-conforms-replacement-haserror-isopenedanyobject-replacement-replacement-is-gener.swift (fixed by @slavapestov in #6484)
* 28490-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28491-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28493-resolveidenttypecomponent-swift-typechecker-swift-declcontext-llvm-arrayref-swif.swift (fixed, see #6505)
* 28494-conforms-type-does-not-conform-to-protocol.swift (fixed by @slavapestov in #6484)
* 28497-unreachable-executed-at-swift-lib-ast-type-cpp-294.swift (fixed by @xedin in #6497)
* 28499-start-isvalid-end-isvalid-start-and-end-should-either-both-be-valid-or-both-be-i.swift (fixed by @CodaFi in #6504)
* 28501-haderror-m-is-sourcefile-m-get-sourcefile-aststage-sourcefile-typechecked-overlo.swift (fixed by @xedin in #6489)
* 28503-iscomplete-missing-inherited-mapping-in-conformance.swift (fixed by @slavapestov in #6484)
* 28504-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in #6481)
* 28505-failed-call-arguments-did-not-match-up.swift (fixed by @xedin in #6492)
* 28507-swift-typerepr-getsourcerange-const.swift (fixed by @slavapestov in #6484)
* 28512-anonymous-namespace-traversal-visit-swift-typerepr.swift (fixed by @slavapestov in #6484)
* 28515-getmutableaddressor.swift (fixed by @slavapestov in #6484)
* 28517-getstoragekind-stored-storagekind-already-set.swift (fixed by @slavapestov in #6484)
* 28518-anonymous-namespace-verifier-walktoexprpost-swift-expr.swift (fixed by @xedin in #6489)
* 28520-args-size-fnref-getnumargumentsforfullapply-partial-application-was-throwing.swift (fixed by @xedin in #6501)
* 28527-e-gettype-isassignabletype-setting-access-kind-on-non-l-value.swift (fixed by @slavapestov in #6484)
* 28535-unreachable-executed-at-swift-lib-parse-parsedecl-cpp-610.swift (fixed by @slavapestov in #6484)
* 28537-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28541-anonymous-namespace-verifier-walktodeclpost-swift-decl.swift (fixed by @slavapestov in #6484)
* 28544-swift-type-transform-llvm-function-ref-swift-type-swift-type-const.swift (fixed by @xedin in #6491)
* 28546-anonymous-namespace-verifier-walktodeclpost-swift-decl.swift (fixed by @slavapestov in #6484)
* 28549-swift-typebase-getdesugaredtype.swift (fixed by @slavapestov in #6484)
* 28554-reftype-hastypeparameter-cannot-have-a-dependent-type-here.swift (fixed by @slavapestov in #6484)
* 28559-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28560-unreachable-executed-at-swift-lib-ast-type-cpp-1104.swift (fixed by @xedin in #6481)
* 28561-input-hastypevariable-output-hastypevariable.swift (fixed by @slavapestov in #6484)
* 28562-swift-typebase-getcanonicaltype.swift (fixed by @xedin in #6481)
* 28567-anonymous-namespace-verifier-verifychecked-swift-type-llvm-smallptrset-swift-arc.swift (fixed by @xedin in #6481)
* 28568-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28569-swift-declcontext-getastypeortypeextensioncontext-const.swift (fixed by @slavapestov in #6484)
* 28576-anonymous-namespace-findcapturedvars-checktype-swift-type-swift-sourceloc.swift (fixed by @xedin in #6481)
* 28577-isa-x-val-cast-ty-argument-of-incompatible-type.swift (fixed by @slavapestov in #6484)
* 28578-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28580-hastype-e-expected-type-to-have-been-set.swift (fixed by @slavapestov in #6484)
* 28581-anonymous-namespace-findcapturedvars-checktype-swift-type-swift-sourceloc.swift (fixed by @xedin in #6481)
* 28582-result-case-not-implemented.swift (fixed by @xedin in #6481)
* 28583-unreachable-executed-at-swift-lib-ast-type-cpp-1098.swift (fixed by @xedin in #6481)
* 28584-loc-isvalid.swift (fixed by @slavapestov in #6484)
* 28585-anonymous-namespace-verifier-walktostmtpost-swift-stmt.swift (fixed by @CodaFi in #6504)
* 28586-objectty-is-lvaluetype-objectty-is-inouttype-cannot-have-inout-or-lvalue-wrapped.swift (fixed by @xedin in #6511)
* 28587-child-source-range-not-contained-within-its-parent-sequence-expr-type-null.swift (fixed by @CodaFi in #6504)
* 28588-unreachable-executed-at-swift-lib-sema-csapply-cpp-5770.swift (fixed by @xedin in #6515)
* 28592-objectty-is-lvaluetype-objectty-is-inouttype-cannot-have-inout-or-lvalue-wrapped.swift (fixed by @xedin in #6510)
* 28594-anonymous-namespace-verifier-verifychecked-swift-vardecl.swift (fixed by @xedin in #6489)

This one should probably be reported upstream to LLVM:
* [28510-size-t-llvm-sys-locale-columnwidth-i-gettext-i-gettext-size.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28510-size-t-llvm-sys-locale-columnwidth-i-gettext-i-gettext-size.swift)

Crashers with a known common root cause that likely will be fixed by @rudkx's upcoming improvements to the constraint system (see discussion in #6484):
* [28442-swift-typebase-getrvaluetype.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28442-swift-typebase-getrvaluetype.swift)
* [28474-unreachable-executed-at-swift-lib-ast-type-cpp-1325.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28474-unreachable-executed-at-swift-lib-ast-type-cpp-1325.swift)
* [28479-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28479-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift)
* [28543-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28543-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift)
* [28550-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28550-unreachable-executed-at-swift-include-swift-ast-typevisitor-h-39.swift)
* [28555-unreachable-executed-at-swift-lib-ast-type-cpp-1318.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28555-unreachable-executed-at-swift-lib-ast-type-cpp-1318.swift)
* [28590-exprtypes-e-isequal-e-gettype-expected-type-in-map-to-be-the-same-type-in-expres.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28590-exprtypes-e-isequal-e-gettype-expected-type-in-map-to-be-the-same-type-in-expres.swift)
* [28593-unreachable-executed-at-swift-lib-ast-type-cpp-3771.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28593-unreachable-executed-at-swift-lib-ast-type-cpp-3771.swift)

Crashers with a known common root cause identified by @slavapestov (see discussion in #6484, ping @slavapestov if you want to help with these :-)):
* [28350-swift-typechecker-performtypocorrection.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28350-swift-typechecker-performtypocorrection.swift)
* [28389-swift-lookupvisibledecls.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28389-swift-lookupvisibledecls.swift)
* [28420-swift-lowering-emitconditionalcheckedcast.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28420-swift-lowering-emitconditionalcheckedcast.swift)
* [28482-hasaccessibility-accessibility-not-computed-yet.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28482-hasaccessibility-accessibility-not-computed-yet.swift)
* [28508-unreachable-executed-at-swift-lib-sema-csgen-cpp-2656.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28508-unreachable-executed-at-swift-lib-sema-csgen-cpp-2656.swift)
* [28509-allowoverwrite-e-haslvalueaccesskind-l-value-access-kind-has-already-been-set.swift](https://github.com/apple/swift/blob/master/validation-test/compiler_crashers/28509-allowoverwrite-e-haslvalueaccesskind-l-value-access-kind-has-already-been-set.swift)

rudkx commented:
@practicalswift I'm still working through issues to get my changes working, but can confirm that at least `28590-exprtypes-e-isequal-e-gettype-expected-type-in-map-to-be-the-same-type-in-expres.swift` is fixed at this point. Hopefully I can wrap this stuff wrapped up soon.

xedin commented:
@rudkx \o/

practicalswift commented:
@rudkx That's excellent! 👍

slavapestov commented:
@swift-ci Please test

jtbandes commented:
@slavapestov What happened to this branch?

slavapestov commented:
@jtbandes Everything got merged a while ago :)
